usecases/allowed_networks_usecase.go

```diff
@@ -0,0 +1,113 @@
+package usecases
+
+import (
+	"context"
+	"net"
+	"net/http"
+
+	"github.com/checkmarble/marble-backend/repositories"
+	"github.com/checkmarble/marble-backend/usecases/executor_factory"
+	"github.com/checkmarble/marble-backend/utils"
+	"github.com/gin-gonic/gin"
+)
+
+type AllowedNetworksUse int
+
+const (
+	AllowedNetworksLogin AllowedNetworksUse = iota
+	AllowedNetworksOther
+)
+
+type ipWhitelistRepository interface {
+	GetOrganizationAllowedNetworks(ctx context.Context, exec repositories.Executor, orgId string) ([]net.IPNet, error)
+}
+
+type AllowedNetworksUsecase struct {
+	executorFactory executor_factory.ExecutorFactory
+	repository      ipWhitelistRepository
+}
+
+func (uc AllowedNetworksUsecase) Guard(use AllowedNetworksUse) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		var buf *utils.BufferResponseWriter
+
+		// If the guard is used for initial login, we actually need to perform
+		// the login in order to know the user's organization and fetch its
+		// whitelist.
+		//
+		// We use a dummy writer in order to be able to intercept the reponse
+		// after it has been written.
+		if use == AllowedNetworksLogin {
+			buf = utils.NewBufferResponseWriter(c)
+
+			c.Next()
+		}
+
+		// We abort if the login was unsuccessful.
+		if c.IsAborted() {
+			return
+		}
+
+		ctx := c.Request.Context()
+		logger := utils.LoggerFromContext(ctx)
+		creds, ok := utils.CredentialsFromCtx(ctx)
+
+		if !ok {
+			logger.WarnContext(ctx, "a request with no credentials entered the IP whitelisting middleware")
+			c.AbortWithStatus(http.StatusUnauthorized)
+			return
+		}
+
+		clientIp := net.ParseIP(c.Request.Header.Get("x-real-ip"))
+
+		// Self-hosted users might not have set the header on their reverse
+		// proxy, so we fail open if it is not set.
+		if clientIp == nil {
+			if use == AllowedNetworksLogin {
+				buf.Restore(c)
+			}
+
+			return
+		}
+
+		// TODO: we need to discuss about this: even when whitelisting is
```

Comment:
This is an important one to discuss.

I want, before letting the admin configure its allowed networks, to check that they are currently on one of those networks (so they don't cut themselves off). To do that, I would need to rely on the "real IP" header from being present.

That implies the user's infrastructure was properly set up to add that header **and** to block values sent by the client.

The line being at this location has an impact on logs: the content of that header will be the IP address written to the logs, so if infrastructure does not filter client-sent data, it will be spoofable.

---

One solution would be to gate IP whitelisting being (yet another...) configuration flag, which would solve the issue but it not amazing UX-wise.